tests/ArrayAssignmentTest.php

```diff
@@ -2335,9 +2350,10 @@ function getList(array $list): array {
 
                         return $list;
                     }',
-                'error_message' => 'LessSpecificReturnStatement',
+                'error_message' => 'InvalidReturnStatement',
```

Comment:
I think this should still be a `LessSpecificReturnStatement` — we're returning an `array<int, string>`, which is a supertype of `list<string>`.

Reply:
Actually this test was skipped because it previously relied on a incorrect ref-out type for `array_pop` which was fixed in this PR.
Now `array_pop` on a list correctly keeps the listness of the array (when previously it was turned into an array), but I couldn't find an easy way to make this complex example work with the new technically correct behavior of array_pop:

```
[daniil@daniil-arch psalm]$ php a.php ; cat a.php
array(3) {
  [0]=>
  string(1) "0"
  [1]=>
  string(1) "1"
  [2]=>
  string(1) "2"
}
array(2) {
  [0]=>
  string(2) "0!"
  [1]=>
  string(1) "1"
}
array(2) {
  [0]=>
  string(2) "0!"
  [1]=>
  string(2) "1!"
}
array(2) {
  [0]=>
  string(2) "0!"
  [2]=>
  string(2) "2!"
}
<?php
                    /**
                     * @param list<string> $list
                     * @return list<string>
                     */
                    function getList(array $list): array {
                        foreach ($list as $key => $value) {
                            var_dump($list);
                            array_pop($list);

                            $list[$key] = $value . "!";
                        }

                        return $list;
                    }
                    var_dump(getList(["0", "1", "2"]));
```

Reply:
Here array_pop pops the 1! value at key 1, still keeping listness, but the next line sets a 2! value at key 2, breaking listness: I couldn't think of an easy way to make Psalm correctly infer the non-listness of the return type.

Reply:
Will try fixing this really nasty edge case in a future PR, if at all possible: right now the dependent list key logic works exactly the same as before, the only reason why this test is broken is due to the new (correct) behavior of array_pop.